src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/BoundConstants.cs

```diff
@@ -197,10 +197,6 @@ private void EmitConstantFromArray(LambdaCompiler lc, object value, Type type)
             {
                 lc.IL.Emit(OpCodes.Unbox_Any, type);
             }
-            else if (type != typeof(object))
-            {
-                lc.IL.Emit(OpCodes.Castclass, type);
-            }
```

Comment:
It will always be something where the `castclass` will push the same thing it popped without throwing, or else the expression producing this would have thrown in the factory.

E.g. one of the test cases has a constant value `new D(5)` and type `I` where `I` is an interface it implements. When it's `ldelem.ref` the `object` storing that `D(5)` value and does `castclass I` it's guaranteed to leave the stack as is and can be ignored in a similar way to how the C# compiler would ignore an upcast or cast-to-own-type on a reference-type constant. At the level of the expression itself, this *is* an up-cast.

To have any other behaviour we'd have to have e.g. `Expression.Constant("string", typeof(Uri))` being compiled, but that factory would always throw on that call.